pkg/pmem-csi-driver/parameters.go

```diff
@@ -0,0 +1,276 @@
+/*
+Copyright 2019 Intel Corporation
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package pmemcsidriver
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/api/resource"
+)
+
+type persistencyModel string
+type parameterOrigin int
+
+// Beware of API and backwards-compatibility breaking when changing these string constants!
+const (
+	parameterCacheSize        = "cacheSize"
+	parameterEraseAfter       = "eraseafter"
+	parameterName             = "name"
+	parameterPersistencyModel = "persistencyModel"
+	parameterVolumeID         = "_id"
+	parameterSize             = "size"
+
+	// Kubernetes v1.16+ adds this key to NodePublishRequest.VolumeContext
+	// while provisioning ephemeral volume.
+	parameterEphemeral = "csi.storage.k8s.io/ephemeral"
+
+	// Additional, unknown parameters that are okay.
+	parameterPodInfoPrefix = "csi.storage.k8s.io/"
+
+	persistencyNormal    persistencyModel = "normal" // In releases <= 0.6.x this was called "none", but not documented.
+	persistencyCache     persistencyModel = "cache"
+	persistencyEphemeral persistencyModel = "ephemeral" // only used internally
+
+	// Parameters are from storage class in controller CreateVolume.
+	createVolumeParameters parameterOrigin = iota
+	// Parameters are from master in node CreateVolume.
+	createVolumeInternalParameters
+	// Parameters are for an ephemeral volume in NodePublishVolume.
+	ephemeralVolumeParameters
+	// Parameters are for a persistent volume in NodePublishVolume.
+	persistentVolumeParameters
```

Comment:
Origin of this(`persistentVolumeParameters`) type is defined here but not used anywhere in the code. I think we can remove.

Reply:
It's actually the opposite: the code which should have used it (the persistent volume code path in `NodePublishVolume`) didn't validate the content of the volume context. That's bad because robust code should reject unknown and/or unexpected input. I've fixed and at the same time added the volume name to the volume context because that is actually more useful than replicating the `CreateVolume` parameters.